pkg/cmds/run_hook.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		return "", err
+	}
+
+	// AppBinding should have a Service in ClientConfig field. This service selects the app pod. We will execute the hooks in the app pod.
+	if appbinding.Spec.ClientConfig.Service != nil {
+		// there should be an endpoint with same name as the service which contains the name of the selected pods.
+		endPoint, err := opt.kubeClient.CoreV1().Endpoints(opt.namespace).Get(appbinding.Spec.ClientConfig.Service.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+
+		for _, subSets := range endPoint.Subsets {
+			// get pod from the ready addresses
+			for _, readyAddrs := range subSets.Addresses {
+				if readyAddrs.TargetRef != nil && readyAddrs.TargetRef.Kind == apis.KindPod {
+					return readyAddrs.TargetRef.Name, nil
+				}
+			}
+			// no pod found in ready addresses. now try in not ready addresses.
+			for _, notReadyAddrs := range subSets.NotReadyAddresses {
+				if notReadyAddrs.TargetRef != nil && notReadyAddrs.TargetRef.Kind == apis.KindPod {
+					return notReadyAddrs.TargetRef.Name, nil
+				}
+			}
+		}
+	}
+	return "", fmt.Errorf("no pod found for AppBinding %s/%s", opt.namespace, appbindingName)
+}
+
+func (opt *hookOptions) handlePreTaskHookFailure(hookErr error) error {
+	statusOpt := status.UpdateStatusOptions{
+		Config:      opt.config,
+		KubeClient:  opt.kubeClient,
+		StashClient: opt.stashClient,
+		Namespace:   opt.namespace,
+		Metrics:     opt.metricOpts,
+	}
+	if opt.hookType == apis.PreBackupHook {
+		backupOutput := &restic.BackupOutput{
+			HostBackupStats: []v1beta1.HostBackupStats{
+				{
+					Hostname: opt.hostname,
+					Phase:    v1beta1.HostBackupFailed,
+					Error:    hookErr.Error(),
+				},
+			},
+		}
+		statusOpt.BackupSession = opt.backupSessionName
+
+		err := statusOpt.UpdatePostBackupStatus(backupOutput)
+		if err != nil {
+			hookErr = errors.NewAggregate([]error{hookErr, err})
+		}
+	} else { // otherwise it is postRestore hook
+		restoreOutput := &restic.RestoreOutput{
+			HostRestoreStats: []v1beta1.HostRestoreStats{
+				{
+					Hostname: opt.hostname,
+					Phase:    v1beta1.HostRestoreFailed,
+					Error:    hookErr.Error(),
+				},
+			},
+		}
+		statusOpt.RestoreSession = opt.restoreSessionName
+
+		err := statusOpt.UpdatePostRestoreStatus(restoreOutput)
+		if err != nil {
+			hookErr = errors.NewAggregate([]error{hookErr, err})
+		}
+	}
+	// return error so that the container fail
+	return hookErr
+}
+
+func (opt *hookOptions) handlePostTaskHookFailure(hookErr error) error {
+	if opt.hookType == apis.PostBackupHook {
+		backupOutput := &restic.BackupOutput{
+			HostBackupStats: []v1beta1.HostBackupStats{
+				{
+					Hostname: opt.hostname,
+					Phase:    v1beta1.HostBackupFailed,
+					Error:    hookErr.Error(),
+				},
+			},
+		}
+
+		err := backupOutput.WriteOutput(filepath.Join(opt.outputDir, restic.DefaultOutputFileName))
```

Comment:
How are we using this?

Reply:
https://github.com/stashed/stash/blob/f49995adbf0512d3043936b71b3a39957a739f3a/pkg/cmds/run_hook.go#L89